CFE-543 : Apply user defined tags to all Azure resources created by OpenShift (#1191)
openshift/enhancements: enhancements/api-review/azure_user_defined_tags.md

- Add `userTags` field to `platform.azure` of the `installconfigs.install.openshift.io`
```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: installconfigs.install.openshift.io
spec:
  versions:
  - name: v1
    schema:
      openAPIV3Schema:
        properties:
          platform:
            properties:
              azure:
                properties:
                  userTags:
                    additionalProperties:
                      type: string
                    description: UserTags additional keys and values that the installer
                      will add as tags to all resources that it creates. Resources
                      created by the cluster itself may not include these tags.
                  type: object
```

- Add `resourceTags` field to `platformStatus.status.azure` 
  of the `infrastructure.config.openshift.io`
```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: infrastructures.config.openshift.io
status:
  versions:
  - name: v1
    schema:
      openAPIV3Schema:
        properties:
          status:
            properties:

>               platformStatus:

Reviewer: The AWS equivalent here is in the status, however, that has caused issues as we now have requests for users to be able to update the tags. I think this should go into the spec, though you'll need to explore the implications of that change as it's different from AWS. There is an enhancement somewhere that discusses the AWS change

Author: Thank you for the comment!! 
Yeah, but for Azure and GCP ask in RFE-2017 is for tagging of resources created during Day-0 operation for now. So thought to do it, as it was done for AWS, as including it in Spec, will give out the idea it can edited. Should we add it in spec, to avoid issues we had with AWS and document saying updating userTags will have no effect?

Reviewer 2: > we now have requests for users to be able to update the tags

Do we?  The last conversations i had w/ anyone one the subject, we concluded that create-time only was fine, at least for user tags.  Are you referring to user provided tags or something else?

Reviewer: It's possible that I'm out of date here, as far as I knew there was an effort earlier this year to support updating tags on AWS. If the conclusion of that was that we aren't going to do it and we are never going to do it, and we are going to apply the same conclusion here, then it seems reasonable to follow the pattern of putting this in the status. 

Do we have any record of that conversation?

Reviewer: > Should we add it in spec to avoid issues we had with AWS and document saying updating userTags will have no effect?

We could put it in the spec and use a webhook to make it immutable, this would provide the same guarantees (technically better) as putting it in status but would allow us to change it later, though removing immutability from an API could be considered a breaking change 🤔

Author: If CEL is used, we should have tags in spec and status, and validate if the tags and the tag count matches. And status will be updated regardless of whether all operators successfully applied the tags on the resources it manages, as in the current proposal.
But CEL is supported in 1.25 beta version, can we use the beta version?

Reviewer: Beta means it's on by default, so we are using CEL from openshift 4.12 onwards

see-also:
  - "enhancements/api-review/custom-tags-aws.md"
replaces:
  - N/A
superseded-by:
  - N/A
---

# Apply user defined tags to all Azure resources created by OpenShift

## Summary

This enhancement describes the proposal to allow an administrator of Openshift to 
have the ability to apply user defined tags to those resources created by Openshift 
in Azure.

## Motivation

Motivations include but are not limited to:

- Allow admin, compliance, and security teams to keep track of assets and objects 
  created by OpenShift in Azure.

### User Stories

- As an openshift administrator, I want to have tags added to all resources created 
  in Azure by Openshift.
- As an openshift administrator, I want to restrict access granted to Openshift specific account.

### Goals

- The administrator or service (in the case of Managed OpenShift) installing OpenShift 
  can configure allowed number of user-defined tags in the Openshift installer generated
  install config, which is referred and applied by installer and the in-cluster operators
  on the the Azure resources during creation.
- Tags must be applied at creation time, in an atomic operation. It isn't acceptable 

>   to create an object and, after some period of time, apply the tags post-creation.

Reviewer 3: Is there a list of objects where this cannot be met?  I haven't looked at Azure specifically but there are some resources in other cloud providers that cannot be tagged atomically, it requires an update to the resource after creation.

Author: The Azure resources which are created by installer using terraform API's, all have support for adding tags during creation. Didn't come across any resource, those created by installer, requiring additional request for adding tags.

### Non-Goals

> - Management of resource tags post creation of cluster is out of scope.

Reviewer 3: I think the intent of this is the set of user tags can be set at cluster creation and are immutable, the tags are set on resources created by the cluster day-2.  Is that correct?

Author: Yes, correct, meant to state, updating or deleting tags added during cluster creation is not supported.

## Proposal

A tag of the form `kubernetes.io/cluster/<cluster_name>:owned` will be added to every
resource created by Openshift to enable administrator to differentiate the resources
created for Openshift cluster. An administrator is not allowed to add or modify the tag 
having the prefix `kubernetes.io` or `openshift.io` in the name. 

New `userTags` field will be added to `platform.azure` of install-config for the user 
to define the tags to be added to the resources created by installer and in-cluster operators.

If `platform.azure.userTags` of install-config has any tag defined same will be added 
to all the azure resources created by Openshift except, when the tag validation fails 
due to any of below conditions
1. A tag name can have a maximum of 128 characters.
    - Tag name has a limit of 512 characters for all resources except for 
    storage accounts, which has a limit of 128 characters and hence tag name 
    length is restricted to 128 characters on every resource required by Openshift.
2. A tag name cannot contain `<, >, %, &, \, ?, /, #, :, whitespace` characters and 

>    should not start with a number.

Reviewer 3: ```suggestion
   must not start with a number.
```

Author: comment incorporated

field to cloud resources, they must never remove tags from the existing underlying cloud 
resource even if the tags are removed from this field(despite it being immutable).

If the userTags field is changed post-install, there is no guarantee about how an 
in-cluster operator will respond to the change. Some operators may reconcile the 
change and change tags on the Azure resource. Some operators may ignore the change. 
However, if tags are removed from userTags, the tag will not be removed from the 
Azure resource.

### Workflow Description

- An Openshift administrator requests to add required tags to all Azure resources 
  created by Openshift by adding it in `.platform.azure.userTags`
- openshift installer validates the tags defined in `.platform.azure.userTags` and 
  adds these tags to all resources created during installation and also updates 
  `.spec.platformSpec.azure.resourceTags` of the `infrastructure.config.openshift.io`
- In cluster operators refers `.spec.platformSpec.azure.resourceTags` of the 
  `infrastructure.config.openshift.io` to add tags to resources created later.
- An Openshift administrator can modify existing tags or add new tags by updating 

>   `.spec.platformSpec.azure.resourceTags` field in the `infrastructure.config.openshift.io`.

Reviewer 3: Earlier it is stated "If the userTags field is changed post-install, there is no guarantee about how an in-cluster operator will respond to the change." and the Non-goals imply modification of userTags to be out of scope.  Will there be any restriction on this property to limit unexpected behavior, so that updates require intentional removal of an admission controller?

Author: Initially when creating this proposal, had covered update and delete of tags too, but it's out of scope with the current requirement. I have missed updating above section with the same.

  install config, which is referred and applied by the installer and the in-cluster operators
  on the the Azure resources during cluster creation.
- Tags must be applied at creation time, in an atomic operation. It isn't acceptable 
  to create an object and to apply tags post cluster creation.

### Non-Goals

- Management(update/delete) of resource tags post creation of cluster is out of scope.

## Proposal

A tag of the form `kubernetes.io/cluster/<cluster_name>:owned` will be added to every
resource created by Openshift to enable administrator to differentiate the resources
created for Openshift cluster. An administrator is not allowed to add or modify the tag 
having the prefix `kubernetes.io` or `openshift.io` in the name. The same tag can be 
found applied to other cloud platform resources which supports tagging for ex: AWS.

New `userTags` field will be added to `platform.azure` of install-config for the user 
to define the tags to be added to the resources created by installer and in-cluster operators.

If `platform.azure.userTags` of install-config has any tag defined same will be added 
to all the azure resources created by Openshift except, when the tag validation fail

> to meet any of the below conditions

Reviewer 4: Minor nitpick: this sounds like you're saying the installer will take the list and apply all of the good labels while ignoring the bad ones. In this case and in general, installer validation will take user input and validate it; if the config fails validation the installer will return an error until the bad input is fixed.

Author: Sorry, I meant to convey, if any tags are configured it will be applied to the resources and the action is not controlled by any feature flag like the experimentalPropagateUserTags(PropagateUserTags) in AWS. And tags will be applied only if all the user defined tags meet the required conditions.

> 1. A tag name can have a maximum of 128 characters.

Reviewer: Do we have any limit on the name of clusters? Could the combination of a long cluster name and the `kubernetes.io/cluster/` prefix cause us to blow this limit? And if so, how do we handle that?

Author: Cluster name [generated](https://github.com/openshift/api/blob/28228c0f01ab20e1d49d8369157fb171ad53e102/config/v1/types_infrastructure.go#L52-L55) by appending a five characters length random string(+1 for `-` used as delimiter) to the user inputted name is restricted to 27 characters, which is within the limit and is not a problem I think.
/cc @patrickdillon Could you please confirm, if my understanding is correct?

Reviewer: If we can verify that's correct, it might be worth noting that somewhere within the enhancement to prevent others asking the same question

Author: I did verify deploying a cluster on Azure. The first catch is when user inputs the cluster name during install-config creation an error is thrown if name has more than 54 characters. And later during the generation of InfraID, if the provided cluster name has more than 21 characters, it is trimmed to 21 characters and a 5 character random string is appended with `-` as a delimiter. The maximum length of a default tag will come to 49 characters.

Reviewer: Ok, please make sure this gets noted in the enhancement somewhere, useful breadcrumbs

Author: Content [added](https://github.com/openshift/enhancements/pull/1191/commits/5cd152284b32fb3ca953b55b031a7c52c9c9e078) to the enhancement.

  adds these tags to all resources created during installation and also updates 
  `.status.platformStatus.azure.resourceTags` of the `infrastructure.config.openshift.io`
- In cluster operators refers `.status.platformStatus.azure.resourceTags` of the 
  `infrastructure.config.openshift.io` to add tags to resources created later.

#### Variation [optional]

### API Extensions
Enhancement requires below modifications to the mentioned CRDs
- Add `userTags` field to `platform.azure` of the `installconfigs.install.openshift.io`
```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: installconfigs.install.openshift.io
spec:
  versions:
  - name: v1
    schema:
      openAPIV3Schema:
        properties:
          platform:
            properties:
              azure:
                properties:
                  userTags:
                    additionalProperties:
                      type: string
                    description: UserTags additional keys and values that the installer
                      will add as tags to all resources that it creates. Resources
                      created by the cluster itself may not include these tags.

>                   type: object

Reviewer 4: not sure if I'm reading this right, and which type is referring to which, but i would expect `userTags` to be the equivalent of a map in go

Author: Yes, it will be of type `map[string]string`

Author: userTags in installconfig CR is defined as `map[string]string` type and the resourceTags in infrastructure CR is defined as slice of struct containing key and value as members of type string.

```

### Implementation Details/Notes/Constraints [optional]
Add a new field `resourceTags` to `.status.platformStatus.azure` of the 
`infrastructure.config.openshift.io` type. Tags included in the `resourceTags` field 
will be applied to new resources created for the cluster by the in-cluster operators.

The `resourceTags` field in `status.platformStatus.azure` of `infrastructure.config.openshift.io`
will be populated by the installer using the entries from `platform.azure.userTags` field of `install-config`.

`status.platformStatus.azure` of `infrastructure.config.openshift.io` is immutable.

All operators that create Azure resources will apply these tags to all Azure 
resources they create. List of in-cluster operators managing cloud resources
could vary across platform types, example for AWS there are additional operators
like aws-ebs-csi-driver-operator, aws-efs-csi-driver-operator to manage specific
resources.

| Operator | Resources created by the operator |
| -------- | ----------------------------- |
| cloud-network-config-controller | Private IP address |
| cluster-image-registry-operator | Storage Account |
| cluster-ingress-operator | Load Balancer, DNS records |

> | cloud-credential-operator | IAM roles and policies |

Reviewer 5: Enhancement looks good wrt CCO which does not presently have Mint mode for Azure due to the deprecation of the graph API (see https://github.com/openshift/cloud-credential-operator/pull/433) and would not create resources. However if CCO were to create resources, CCO could check the infrastructure object to get the list of tags.

There is an upcoming request for short-lived credential support on Azure and the IAM resources would be created by CCO's command-line utility (ccoctl) ahead of installation (see https://github.com/openshift/enhancements/pull/1301). Is there a need to provide a way to tag these resources? CCO would not have access to the install-config when creating this infrastructure.

Author: My bad, I missed this while doing code walk through for the enhancement. CCO reconciles just the credentials secret but does not create or update any credentials.

/cc @TrilokGeer
Customer feedback was to have all possible resources, created for OCP, to be tagged. But since resources are created before cluster creation, should ccoctl provide options for tags with azure identity resource creation related sub-commands?

Reviewer 6: @abutcher @bharath-b-rh , looking at https://github.com/openshift/enhancements/pull/1301, there'd be a need to have capabilities to tag the managed identities created by command-line utility (ccoctl). Assuming the changes in ccoctl are specific to OpenShift, it is suggested to have default tag ("kubernetes.io/cluster/${var.cluster_id}" = "owned") added by ccoctl. Though there are some scenarios which have limitations, it is detailed at https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/tag-support#microsoftmanagedidentity.

| machine-api-provider-azure | Application Security Group, Availability Set, Group, Load Balancer, Public IP Address, Route, Network Security Group, Virtual Machine Extension, Virtual Interface, Virtual Machine, Virtual Network. |

Below list of terraform Azure APIs to create resources should be updated to add user
defined tags and as well the openshift default tag in the installer component.

> `azurerm_resource_group, azurerm_image, azurerm_lb, azurerm_network_security_group, azurerm_storage_account, azurerm_user_assigned_identity, azurerm_virtual_network, azurerm_linux_virtual_machine, azurerm_network_interface, azurerm_dns_cname_record`

Reviewer 4: I'm not sure I understand the significance of this list. I'm reading this as: this is a list of resources created by the installer that don't have tags and tags will be added to them. Is that correct?

Author: Yes, those are the terraform APIs for respective azure resources which needs the change to apply the tags.

superseded-by:
  - N/A
---

# Apply user defined tags to all Azure resources created by OpenShift

## Summary

This enhancement describes the proposal to allow an administrator of Openshift to 
have the ability to apply user defined tags to those resources created by Openshift 
in Azure.

## Motivation

Motivations include but are not limited to:

- Allow admin, compliance, and security teams to keep track of assets and objects 
  created by OpenShift in Azure.

### User Stories

- As an openshift administrator, I want to have tags added to all resources created 
  in Azure by Openshift.
- As an openshift administrator, I want to restrict access granted to Openshift 
  specific account.

### Goals

- The administrator or service (in the case of Managed OpenShift) installing OpenShift 
  can configure allowed number of user-defined tags in the Openshift installer generated
  install config, which is referred and applied by the installer and the in-cluster operators
  on the the Azure resources during cluster creation.
- Tags must be applied at creation time, in an atomic operation. It isn't acceptable 
  to create an object and to apply tags post cluster creation.

### Non-Goals

> - Management(update/delete) of resource tags post creation of cluster is out of scope.

Reviewer 2: last i knew we'd killed those efforts, are they back?

> - Management(update/delete) of resource tags post creation of cluster is out of scope.

Reviewer 2: at the time we dropped them, we surveyed the customers who wanted AWS tagging in general, and they made it clear all they needed was the ability to tag resources at creation time.

so we dropped it:
1) because no one needed it (so why bother w/ the complexity/effort?)
2) updating tags gets into lots of fun questions about how to reconcile conflicts w// manual edits, what to do about deletions, what to do if you exceed the tag count limits, etc.

That said, i think i saw something recently that AWS tag updating did ultimately land, i don't know/remember what brought it back from the dead.

## Proposal

A tag of the form `kubernetes.io/cluster/<cluster_name>:owned` will be added to every
resource created by Openshift to enable administrator to differentiate the resources
created for Openshift cluster. An administrator is not allowed to add or modify the tag 
having the prefix `kubernetes.io` or `openshift.io` in the name. The same tag can be 
found applied to other cloud platform resources which supports tagging for ex: AWS.

> New `userTags` field will be added to `platform.azure` of install-config for the user 

Author: Yeah, for AWS too, user has to configure through install-config.

| cloud-credential-operator | IAM roles and policies |
| machine-api-provider-azure | Application Security Group, Availability Set, Group, Load Balancer, Public IP Address, Route, Network Security Group, Virtual Machine Extension, Virtual Interface, Virtual Machine, Virtual Network. |

Below list of terraform Azure APIs to create resources should be updated to add user
defined tags and as well the openshift default tag in the installer component.
`azurerm_resource_group, azurerm_image, azurerm_lb, azurerm_network_security_group, azurerm_storage_account, azurerm_user_assigned_identity, azurerm_virtual_network, azurerm_linux_virtual_machine, azurerm_network_interface, azurerm_dns_cname_record`

API update example:
A local variable should be defined, which merges the default tag and the user
defined Azure tags, which should be referred in the Azure resource APIs.
``` terraform
locals {
  tags = merge(
    {
      "kubernetes.io_cluster.${var.cluster_id}" = "owned"
    },
    var.azure_extra_tags,
  )
}

resource "azurerm_resource_group" "main" {
  tags     = local.tags
}
```

#### Caveats
1. User updates or removes Azure resource tags added by Openshift, using an external interface

>    will not be reconciled by the resource owning operator.

Reviewer: Is that true or does it vary per operator? In AWS we found some did put them back (MAPI for example) and others didn't

Author: Yeah, it varies across operator. Few handle it and few do not.

---
title: azure_user_defined_tags
authors:
  - "@bhb"
reviewers:
  - "@patrickdillon"
  - "@sdodson"
  - "@jhixson74"
  - "@JoelSpeed"
  - "@dhellmann"
  - "@jewzaam"
  - "@adambkaplan"
  - "@abhat"
  - "@dmage"
  - "@Miciah"
  - "@abutcher"
  - "@staebler"

>   - "@jerpeter1"

Reviewer 2: this is a lot of people you've asked to review this which:

1) will consume a lot of people's time
2) will make it difficult to reach consensus and move forward

Can you please provide a sentence for each person indicating why you are requesting their review/what aspects of this EP they should be focusing on?

(if someone of these people are only listed for their awareness, i'd just tag/point them to the EP, not list them as reviewers)

Author: Yeah I agree, will comment with a reason for adding as reviewers and will CC others.

see-also:
  - "enhancements/api-review/custom-tags-aws.md"
replaces:
  - N/A
superseded-by:
  - N/A
---

# Apply user defined tags to all Azure resources created by OpenShift

## Summary

This enhancement describes the proposal to allow an administrator of Openshift to 
have the ability to apply user defined tags to those resources created by Openshift 
in Azure.

## Motivation

Motivations include but are not limited to:

- Allow admin, compliance, and security teams to keep track of assets and objects 
  created by OpenShift in Azure.

### User Stories

- As an openshift administrator, I want to have tags added to all resources created 
  in Azure by Openshift, so that I can restrict access granted to an OpenShift specific account.

### Goals

- The administrator or service (in the case of Managed OpenShift) installing OpenShift 
  can configure allowed number of user-defined tags in the Openshift installer generated
  install config, which is referred and applied by the installer and the in-cluster operators
  on the the Azure resources during cluster creation.
- Tags must be applied at creation time, in an atomic operation. It isn't acceptable 

>   to create an object and to apply tags post cluster creation.

Author: Tags must be added to resources as part of creation as a single operation and not anytime later and definitely not after cluster is created. Yes, it would be better to say`post resource creation`.

- As an openshift administrator, I want to have tags added to all resources created 
  in Azure by Openshift.
- As an openshift administrator, I want to restrict access granted to Openshift 
  specific account.

### Goals

- The administrator or service (in the case of Managed OpenShift) installing OpenShift 
  can configure allowed number of user-defined tags in the Openshift installer generated
  install config, which is referred and applied by the installer and the in-cluster operators
  on the the Azure resources during cluster creation.
- Tags must be applied at creation time, in an atomic operation. It isn't acceptable 
  to create an object and to apply tags post cluster creation.

### Non-Goals

- Management(update/delete) of resource tags post creation of cluster is out of scope.

## Proposal

> A tag of the form `kubernetes.io/cluster/<cluster_name>:owned` will be added to every

Reviewer 2: is that whole thing the tag name, with a tag value of `true`?  Or is `owned` the tag value?

Reviewer 2: also why kubernetes and not openshift?  none of this is upstream behavior, right?

what did we name the tag on AWS?

Author: `kubernetes.io/cluster/<cluster_name>` is the tag key and `owned` is the tag value.

I am not sure why `kubernetes` was used in the tag key, but when I checked the code, it's used for other cloud platform resources too like AWS, GCP, IBMCloud, AlibabaCloud.

Reviewer 2: it would be good to mention that this tag name is the same tag name we use for other clouds

to all the azure resources created by Openshift except, when the tag validation fail
to meet any of the below conditions
1. A tag name can have a maximum of 128 characters.
    - Tag name has a limit of 512 characters for all resources except for 
    storage accounts, which has a limit of 128 characters and hence tag name 
    length is restricted to 128 characters on every resource required by Openshift.
2. A tag name cannot contain `<, >, %, &, \, ?, /, #, :, whitespace` characters and 
   must not start with a number.
    - DNS zones, Traffic, Front Door resources does not support tag with spaces, 
    special/unicode characters or starting with number, hence these are added as 
    constraints on every other Azure resource required by Openshift as well.
3. A tag value can have a maximum of 256 characters.
4. A resource, resource-group or subscription, user can configure a maximum of 5 tags
   through Openshift. 
    - Azure supports a maximum of 50 tags except for Automation, Content Delivery Network,
    DNS resources which can have a maximum of 15 tags, hence restricting the number of 
    user defined tags to 5 and 10 for Openshift's internal use, for all the resources 

>     created by Openshift.

Reviewer 2: we recently had to bump up the number of user tags allowed for AWS resources, are we convinced 5 user tags will be enough for azure customers?  Do we really need to reserve 10 for openshift?

Author: Yeah, I guess Azure users too won't be happy with the limit. In Azure, out of the resources OCP requires, just DNS has the limitation and is created by installer and no operator creates new. 
Would it be acceptable to consider the first 10 userTags for DNS resource alone without throwing error, and have the limitation of 40 tags for other resources, with 10 reserved for OCP?
During one of the discussion, what I learnt was, 4 OCP specific tags are added to AWS resources by SRE team, and thought 10 would be a safer value, in case OCP needs to add few more tags in future.

Reviewer 2: > 4 OCP specific tags are added to AWS resources by SRE team

Wouldn't SRE add those via this mechanism(setting the tags in spec.userTags), so they'd be part of the 5 user tag limit, not the 10 openshift tag limit?

Also aren't there really 3 categories of tag here?

1) owned/generated by openshift for its own purposes (you say this is limited to 10)
2) specified by the user in spec.userTags, but applied by openshift controllers to the resource (you say this is limited to 5)
3) specified by the user directly via the cloud interface (this is effectively limited to "whatever is left")

When you say "5 are reserved for user tags" is that category 2, or category 3?

Author: > Wouldn't SRE add those via this mechanism(setting the tags in spec.userTags), so they'd be part of the 5 user tag limit, not the 10 openshift tag limit?

Yeah, with just 5 allowed for user configuration, and 4 is known to be used, it just leaves out 1 for the user and I thought to get the optimal split up during the review.

> When you say "5 are reserved for user tags" is that category 2, or category 3?

5 reserved tags for the user belongs to category 2.

Reviewer 2: We're only using one tag in category 1, right?  (the `owned` tag).  So do we really need to reserve 10 tags for that category?

Author: Category 1 has just one tag. Except for aws-load-balancer-operator, which adds an additional tag, but that's in AWS. Did not find any operator managing Azure resources adding additional tags other than `kubernetes.io.cluster:owned` tag.

Yeah, 10 won't be required, I think 5 for OCP should be sufficient.

Reviewer 2: > Yeah, 10 won't be required, I think 5 for OCP should be sufficient.

so should we flip this, reserve 5 for OCP and limit the user to 10?

Author: Yeah, unless someone objects with a requirement for more tags :)
Changes updated.

Author: As SRE team makes use of userTags for adding tags and tags reserved for OCP are for those added internally by installer & operators like the `kubernetes.io.cluster:owned` default tag, thought 5 should be enough. Do we need to reserve more for OCP's future use?

Reviewer 2: > Can we enumerate the 5 that are in use here because we mention 2 in use for openshift and 4 in use by SRE, which would make 6, so are we sure 5 is enough?

we don't have 5 in use currently, we have 1 in use currently(kubernetes.io.cluster:owned)  (two if you count the aws loadbalancer scenario described above), as i understand it.

So 5 feels like a decent amount of headroom for our needs and we're already getting asked by users to reserve more tags for their usage, so swinging in the balance in that direction seems justified.

Reviewer 4: For AWS, we take the first N (I believe it is 8) lexicographically sorted user tags and apply those to the resources that have lower limits than the cluster limits.

I am not sure if this would be acceptable in the case of "dns resources," but if we think customers are not too concerned about tagging this resource, it seems like it would allow more tags for the cluster.

Reviewer 4: I think we should resolve this thread (i.e. update the enhancement) before moving forward. Are there any issues with an outcome of 50 tags (with a 25/25 user/openshift split)? 

DNS resources are the only cluster resources with the lower limit. Perhaps we could just skip tagging these? If not, we could apply an arbitrary or lexicographically sorted subset of the user provided tags. Keep in mind that both the installer and cluster ingress operator create dns records.

Reviewer 2: can we get a resolution on whether we need to be tagging dns resources or not?

i don't have an issue w/ a 25/25 split for resources that can handle 50.  For resources that can only handle 15, i'd like to see 5(openshift)/10(user) and lexographic sorting so we get a deterministic behavior when truncating the list.  But if someone can get an actual PM answer on whether DNS resources *need* tagging, that would be even better.

Author: @patrickdillon @bparees @JoelSpeed Would it be acceptable if we proceed with 10(user)/5(openshift) split for the tech preview as pointed out by @TrilokGeer.

Reviewer 2: > @patrickdillon @bparees @JoelSpeed Would it be acceptable if we proceed with 10(user)/5(openshift) split for the tech preview as pointed out by @TrilokGeer.

are you suggesting that as the limits for all resources, or the limits for resources that don't allow more tags?

Author: I would like to suggest to have the limitation of max 10 user tags for all the resources.

Reviewer 2: > I would like to suggest to have the limitation of max 10 user tags for all the resources.

since it is easier to raise a limit later than lower it, i guess we can start there and then raise it for the other resources if needed (at the cost of more code complexity and need for clear documentation about the behavior for different resource types).

> Azure resource.

Author: Yes, some operators may add or update tags of a resource during reconciliation but should never remove tags.

- openshift installer validates the tags defined in `.platform.azure.userTags` and 
  adds these tags to all resources created during installation and also updates 
  `.status.platformStatus.azure.resourceTags` of the `infrastructure.config.openshift.io`
- In cluster operators refers `.status.platformStatus.azure.resourceTags` of the 
  `infrastructure.config.openshift.io` to add tags to resources created later.

#### Variation [optional]

### API Extensions
Enhancement requires below modifications to the mentioned CRDs
- Add `userTags` field to `platform.azure` of the `installconfigs.install.openshift.io`
```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: installconfigs.install.openshift.io
spec:
  versions:
  - name: v1
    schema:
      openAPIV3Schema:
        properties:
          platform:
            properties:
              azure:
                properties:
                  userTags:
                    additionalProperties:
                      type: string
                    description: UserTags additional keys and values that the installer
                      will add as tags to all resources that it creates. Resources

>                       created by the cluster itself may not include these tags.

Author: I meant to convey, any resource created by not an openshift core operator or non-openshift operator may or may not include the user defined tags.

>                       created by the cluster itself may not include these tags.

Reviewer 7: I wouldn't have guessed that "created by the cluster" means *not* created by a core operator.  Would "created by user workload" be accurate?

>                       created by the cluster itself may not include these tags.

Author: Yeah, but that wouldn't cover intree cloud controllers created resources.

                          type: object
                          required:
                            - key
                            - value
                          properties:
                            key:
                              description: key is the key of the tag
                              type: string
                              maxLength: 128
                              minLength: 1
                              pattern: ^[a-zA-Z][0-9A-Za-z_.=+-@]+$
                            value:
                              description: value is the value of the tag.
                              type: string
                              maxLength: 256
                              minLength: 1
                              pattern: ^[0-9A-Za-z_.=+-@]+$
```

### Implementation Details/Notes/Constraints [optional]
Add a new field `resourceTags` to `.status.platformStatus.azure` of the 
`infrastructure.config.openshift.io` type. Tags included in the `resourceTags` field 
will be applied to new resources created for the cluster by the in-cluster operators.

The `resourceTags` field in `status.platformStatus.azure` of `infrastructure.config.openshift.io`

> will be populated by the installer using the entries from `platform.azure.userTags` field of `install-config`.

Reviewer 2: does anything else go into resourceTags other than the user tags?  (I guess the `kubernetes.io.cluster:owned` tag would be one)

Do other tags(not from spec.userTags) go into resourceTags in the AWS equivalent?

Author: No, currently just the tags in userTags will be added to resourceTags. Behavior is same for AWS too.

Reviewer 2: so we don't even put `kubernetes.io.cluster:owned` in there?  Is that just hardcoded into the resource controllers to apply?

Author: Yes, resourceTags will have only the user defined tags and the operators will add the default tag.

Reviewer 2: > Yes, resourceTags will have only the user defined tags and the operators will add the default tag.

that seems a little strange to be honest.  I would have expected the user to supply their desired tags in spec, and then status to reflect all the user tags plus all the "default/required by ocp" tags.  And then the operators would look at status and apply that set of tags to the resources they manage.

Having each operator hardcode the default tag seems undesirably distributed(and has potential to be out of sync if one operator doesn't hardcode the right thing, especially if we change/add a new one) instead of centralized.

That said, i realize you're only building on how it works on AWS today, so unless we're going to revisit the AWS version of this implementation, it is probably moot.  I wouldn't want to do something different for Azure than we do for AWS.

Author: Yeah I proposed the implementation the way it was done for AWS. And also I think since userTags cannot be updated it was added in status, as keeping it in spec would give out the idea it can be modified.

Reviewer 2: keeping it in status is fine, but you could(well, could have) still add the additional system-generated tag in status.

Author: Yeah, I agree. Will update the description of resourceTags to state it contains both OCP and user defined tags applied on the resources. 
Also, in a case where operator adds specific tag to a resource it manages, should it also update the resourceTags(it could be added to just one type of resource)?

Reviewer 2: > Will update the description of resourceTags to state it contains both OCP and user defined tags applied on the resources.

so on AWS status won't include it, but on Azure it will?  And then operator/controller implementations will need to know whether or not to add the OCP tags to the list?

I started out asking the question because i wasn't sure what we did in AWS.  Much as i think it would be better to include it all in status, i also think it's probably better to be consistent in how we use status for this across our clouds.

So if we're not going to revisit the AWS status field, we should probably make the Azure one work the same way.

Author: Yeah, but since we are adding support now for azure and gcp, should we make this change in these both and later pick up aws? Otherwise, we will have to revisit all 3 in the future.

`status.platformStatus.azure` of `infrastructure.config.openshift.io` is immutable.

All operators that create Azure resources will apply these tags to all Azure 
resources they create.

| Operator | Resources created by the operator |
| -------- | ----------------------------- |
| cloud-network-config-controller | Private IP address |
| cluster-image-registry-operator | Storage Account |
| cluster-ingress-operator | Load Balancer, DNS records |
| cloud-credential-operator | IAM roles and policies |

> | machine-api-provider-azure | Application Security Group, Availability Set, Group, Load Balancer, Public IP Address, Route, Network Security Group, Virtual Machine Extension, Virtual Interface, Virtual Machine, Virtual Network. |

Reviewer 2: is this the same set of resource owning components as for AWS?  Or is there a delta?

Author: AWS has a few of more operators(ex: aws-ebs-csi-driver-operator, aws-ebs-csi-driver-operator), but for Azure there is no new operator which wasn't there for AWS.

Reviewer 2: you might mention that (it reduces the risk/complexity of adding this feature since it's all areas we've touched before at some level)

Author: comment incorporated.

> | cluster-ingress-operator | Load Balancer, DNS records |

Author: Thank you for confirming. Removed Load Balancer from the list of resources created by ingress operator.

> | cluster-ingress-operator | Load Balancer, DNS records |

Reviewer 7: I had forgotten that we use a unified load balancer on Azure, which means that the ingress operator uses the existing LB.  However, if the ingress operator creates a service with `spec.type: LoadBalancer`, is there a resource for the public IP address or vnet that then needs to be tagged?  (Sorry, I'm a ignorant of the details of how LBs are configured on Azure.)

> | cluster-ingress-operator | Load Balancer, DNS records |

Reviewer 7: For the record, Bharath and I discussed this, and no operator per se creates these resources; rather, the resources are created by the in-tree cloud provider implementation, which is out of scope for this EP.  

@bharath-b-rh, as you pointed out to me, the RFE excludes resources created by the in-tree cloud provider implementation.  It might be worth listing that as an explicit non-goal in the EP.

> | cluster-ingress-operator | Load Balancer, DNS records |

Author: I have included this in the drawbacks section [here](https://github.com/openshift/enhancements/pull/1191/files#diff-07e0630cc477ba74bc69ea6a15ecefd4f5ce32927210e4e7b0b89d423d2193dbR252-R254)

Below list of terraform Azure APIs to create resources should be updated to add user
defined tags and as well the openshift default tag in the installer component.
`azurerm_resource_group, azurerm_image, azurerm_lb, azurerm_network_security_group, azurerm_storage_account, azurerm_user_assigned_identity, azurerm_virtual_network, azurerm_linux_virtual_machine, azurerm_network_interface, azurerm_dns_cname_record`

API update example:
A local variable should be defined, which merges the default tag and the user
defined Azure tags, which should be referred in the Azure resource APIs.
``` terraform
locals {
  tags = merge(
    {
      "kubernetes.io_cluster.${var.cluster_id}" = "owned"
    },
    var.azure_extra_tags,
  )
}

resource "azurerm_resource_group" "main" {
  tags     = local.tags
}
```

#### Caveats
1. User updates or removes Azure resource tags added by Openshift, using an external interface
   will not be reconciled by the resource owning operator.
2. Updating tags of individual resources is not supported and any tag present in 
   `.status.platformStatus.azure.resourceTags` of `infrastructure.config.openshift.io/v1` resource 

>    will result in addding tags to all Openshift managed Azure resources. 

Reviewer 2: ```suggestion
   will result in adding tags to all Openshift managed Azure resources. 
```

Reviewer 2: also not sure this statement is true.  If the resource has been created, and then someone updates spec.usertags, status.usertags will reflect the spec update(right?), but already created resources may or may not have the new tag added (depending on what the component that manages the resource decides to do, and also whether or not it's actually possible to update the tags on an existing azure resource)

Author: What I wanted to suggest was, defining tags for individual resources is not supported, but all the user defined tags will be applied to all the resources that will be created. I will rephrase the statement.

>    may or may not be reconciled by the operator managing the resource.

Author: Yes, changes made to tags of resources using external tools like azure cli, console etc outside of openshift.

>    may or may not be reconciled by the operator managing the resource.

Reviewer 7: Ah, so you mean something like this: "If the user uses an external interface to update or remove tags that OpenShift previously added to a resource, then the tags may or may not be reconciled by the operator managing the resource."

>    may or may not be reconciled by the operator managing the resource.

Author: Yeah correct.

>   operator. In this proposal, the changes proposed and developed will be part of

Author: Yes

  User-defined tags can be updated on the following Azure resources.
    1. Virtual machine resources created for master and worker nodes.
    2. Storage account
    3. NetworkEdge

>     4. Internal Registry

Author: Yes correct, had mentioned in a broader sense as to the OpenShift core areas creating the azure resources, sorry.
I have updated it with the list of azure resources created by Openshift, which can be tagged.

  by it and constraints other resources with the least matching limit as below
    1. Tag names cannot have `microsoft`, `azure`, `windows` prefixes which are
       reserved for Azure use.
    2. An Azure storage account has a limit of 128 characters for the tag name.
    3. An Azure DNS zone or Traffic or Front Door resource tag name cannot have spaces,
       special/unicode characters or start with a number.
    4. An Azure Automation or Content Delivery Network or DNS resource can have a
       maximum of 15 tags.

- Administrator will have to manually perform below tags pertaining actions
    1. removing the undesired tags from the required resources.
    2. update tag values of the required resources.
    3. update tags of the resources which are not managed by an operator.
    4. update tags of the resources for which update logic is not supported by an operator.

## Design Details

### Open Questions

### Test Plan

### Graduation Criteria

#### Dev Preview -> Tech Preview

#### Tech Preview -> GA

#### Removing a deprecated feature

### Upgrade / Downgrade Strategy

On upgrade:
- Cluster operators that update the tags of Azure resources created for cluster 

>   should refer the new fields and take action. 

Reviewer 2: the implication here is that on upgrade we'd expect clusteroperators to read status.resourceTags and then update the tags on existing azure resources?

is that a requirement?  For AWS we only committed to setting the tags on AWS resources at creation time, right?  Updating existing cloud resources w/ new/updated tags was "unspecified" behavior that very much depends on what the underlying resource supports, and what the component managing the resource decided to do.

Author: Yeah, even for Azure, tags will be added to the resources during creation time. But if the operators managing the Azure resources are upgraded, if any tags are present in resourceTags those will be added to resources by such operators which reconcile resource properties.

Reviewer 2: > . But if the operators managing the Azure resources are upgraded, if any tags are present in resourceTags those will be added to resources by such operators which reconcile resource properties.

right, so it's not a requirement, it's unspecified behavior.  Some operators will reconcile the tags, others may not.  I guess that's what you're saying with "Cluster operators that update tags", but i think it's worth calling out the unspecified nature of this behavior more explicitly.

Author: comment incorporated

> - Cluster operators that update the tags of Azure resources created for cluster 
>   should refer the new fields and take action. Any new resource created post-upgrade
>   and the operators managing the resource will add the user defined tags to the 

Reviewer 7: Some copy edits:
```suggestion
- Cluster operators that update the tags of Azure resources created for the cluster 
  should refer to the new fields and take action. For any new resource created post-upgrade,
  the operator managing the resource will add the user-defined tags to the 
```
I was initially confused by the phrasing "operators that update the tags of Azure resources created for cluster"; does that refer to operators that update tags on existing resources when `userTags` is updated?
```suggestion
- Cluster operators that update tags on existing resources when `userTags` is updated
  should refer to the new fields and take action. For any new resource created post-upgrade,
  the operator managing the resource must add the user-defined tags to the 
```

Author: I meant to convey, those operators which manage an azure resource and add tags should refer to the new fields. I will rephrase it to mean adding tags to azure resources.

On downgrade:

> - The status field may remain populated, components may or may not continue 

Reviewer 7: The status field won't exist on downgrade, right?

Author: I thought it may not be cleared. Will it be removed, should I remove the section?

Reviewer 7: The field should be removed if you downgrade to a version of OpenShift that doesn't have it; cluster-version-operator should downgrade cluster-config-operator, which should in turn revert the infrastructures.config.openshift.io CRD to a version that doesn't have the field.  

In my opinion, it makes sense to keep this section to set expectations.  I would expect the consequence of downgrading the cluster to be that downgraded operators might not add tags to new resources and may or may not preserve the tags added prior to the downgrade.

Author: There is no direct fallback/downgrade option, but we have to just upgrade again to a desired version and during this if etcd backup is restored, resources will not be overwritten or recreated. I have updated the section with different upgrade and downgrade scenarios.

  to tag newly created resources with the additional tags depending on regardless of

>   whether given component still has logic to respect the status tags, after the downgrade.

Reviewer 7: Is it "depending on" or is it "regardless of"?

I don't understand this sentence.  Are you saying that components might already have some logic to add tags besides the ones in `userTags`, and those components may apply those tags after downgrade?

Author: It should be "depending on", I will correct it.
Added this statement to be on the safe side, should any operator have the logic to persist userTags in their own object and use it.

  - "@trozet" ## reviewer for cloud-network-config-controller component
approvers:
  - "@jerpeter1" ## approver for CFE
api-approvers:
  - "@JoelSpeed" ## approver for api component
creation-date: 2022-07-12
last-updated: 2022-07-12
tracking-link:
  - https://issues.redhat.com/browse/OCPPLAN-8155
  - https://issues.redhat.com/browse/CORS-2249
see-also:
  - "enhancements/api-review/custom-tags-aws.md"
replaces:
  - N/A
superseded-by:
  - N/A
---

# Apply user defined tags to all Azure resources created by OpenShift

## Summary

This enhancement describes the proposal to allow an administrator of Openshift to 
have the ability to apply user defined tags to those resources created by Openshift 
in Azure.

## Motivation

Motivations include but are not limited to:

- Allow admin, compliance, and security teams to keep track of assets and objects 
  created by OpenShift in Azure.

### User Stories

- As an openshift administrator, I want to have tags added to all resources created 
  in Azure by Openshift, so that I can restrict access granted to an OpenShift specific account.

### Goals

- The administrator or service (in the case of Managed OpenShift) installing OpenShift 

>   can configure allowed number of user-defined tags in the Openshift installer generated

Author: Yes, it's the latter, where an administrator or service can configure tags up to a maximum allowed by the OpenShift..

> - The administrator or service (in the case of Managed OpenShift) installing OpenShift 
>   can configure allowed number of user-defined tags in the Openshift installer generated

Reviewer 7: Maybe this would be clearer?
```suggestion
- The administrator or service (in the case of Managed OpenShift) installing OpenShift 
  can configure a list of up to 10 user-defined tags in the OpenShift installer
```
But if I'm the only person who finds the original wording confusing, then don't worry about it.  * grin *.

>   can configure allowed number of user-defined tags in the Openshift installer generated

Author: comment incorporated

  install config, which is referred and applied by the installer and the in-cluster operators
  on the the Azure resources during cluster creation.
- Tags must be applied at creation time, in an atomic operation. It isn't acceptable 
  to create an object and to apply tags post cluster creation.

### Non-Goals

- Management(update/delete) of resource tags post creation of cluster is out of scope.

## Proposal

> A tag of the form `kubernetes.io_cluster.<cluster_id>:owned` will be added to every

Reviewer: This tag has changed since the initial revision of the enhancement, can you quickly explain why? Is the tag format requirement different in Azure than AWS?

Author: A mistake from me, azure tag names cannot have `/` character and at present tag added to Azure ResourceGroup is of the `kubernetes.io_cluster.<cluster_id>:owned` form.

Reviewer 7: I'm confused.  Does the installer already add the `kubernetes.io_cluster.<cluster_id>:owned` tag?  It looks like it does: https://github.com/openshift/installer/blob/face870099c9408541367651b2d1f45ab7a86a01/pkg/asset/cluster/azure/azure.go#L55

If I understand correctly, it is already an existing, well known tag.  Are you trying to say that (1) the installer will continue to add the `kubernetes.io_cluster.<cluster_id>:owned` tag (i.e. the EP doesn't change this existing behavior), (2) now operators will also add the `kubernetes.io_cluster.<cluster_id>:owned` tag, and (3) the installer and operators will also add the tags from `userTags`?

Reviewer 7: @bharath-b-rh, I'm still confused on this point.

Reviewer 8: @bharath-b-rh - I think the questions in this conversation are still unresolved.

Author: I had missed updating here, installer currently adds just the ocp tag `kubernetes.io_cluster.<cluster_id>:owned` to only `ResourceGroup` resource. With this EP installer and the operators will add ocp and user-defined tags to all the resources created by OpenShift.

resource created by Openshift to enable administrator to differentiate the resources
created for Openshift cluster. An administrator is not allowed to add or modify the tag 
having the prefix `kubernetes.io` or `openshift.io` in the name. The same tag can be 
found applied to other cloud platform resources which supports tagging for ex: AWS.

New `userTags` field will be added to `platform.azure` of install-config for the user 
to define the tags to be added to the resources created by installer and in-cluster operators.

If `platform.azure.userTags` of install-config has any tag defined same will be added 

> to all the azure resources created by Openshift except, when the tag validation fail

Author: Yes, installer validates the tags defined by user in install-config and adds the tags to the resources it creates and as well as in Infrastructure object it generates, only when validation is successful.

> to all the azure resources created by Openshift except, when the tag validation fail

Reviewer 7: If the validation fails, then does the install happen anyway?  I thought the installer would immediately exit out and require the user to provide valid input.

> to all the azure resources created by Openshift except, when the tag validation fail

Reviewer 4: The Installer should block the install until the validation passes. As a general rule, the installer should not change user input; just tell them what is wrong with it and let them fix it.

> to all the azure resources created by Openshift except, when the tag validation fail

Author: Yes, if the validation fails, installer errors out and the user has to retry cluster creation by fixing the tags which is causing the error. The error logs will include the key/value which is invalid.

    },
    var.azure_extra_tags,
  )
}

resource "azurerm_resource_group" "main" {
  tags     = local.tags
}
```

The tag of the form `kubernetes.io_cluster.<cluster_id>:owned` added by openshift, where
cluster_id is a string generated by concatenating user inputted cluster name and a random 
string will be limited to a maximum length of 27 characters by trimming long cluster name 
to 21 characters.

#### Caveats
1. Updating or removing resource tags added by Openshift, using an external interface
   may or may not be reconciled by the operator managing the resource.
2. Updating tags of individual resources is not supported and any tag present in 
   `.status.platformStatus.azure.resourceTags` of `infrastructure.config.openshift.io/v1` resource 
   will result in adding tags to all Openshift managed Azure resources. 

### Risks and Mitigations

### Drawbacks
- User-defined tags cannot be updated on an Azure resource which is not managed by an
  operator. In this proposal, the changes proposed and developed will be part of
  openshift-* namespace. External operators are not in the scope.
  User-defined tags can be updated on the following Azure resources.

>   - ResouceGroup

Reviewer 7: ```suggestion
  - ResourceGroup
```

Author: comment incorporated.

---
title: azure_user_defined_tags
authors:
  - "@bhb"
reviewers:
  - "@patrickdillon" ## reviewer for installer component
  - "@JoelSpeed" ## reviewer for api and machine-api-provider-azure components
  - "@flavianmissi" ## reviewer for cluster-image-registry-operator component
  - "@Miciah" ## reviewer for cluster-ingress-operator component
approvers:
  - "@jerpeter1" ## approver for CFE
api-approvers:
  - "@JoelSpeed" ## approver for api component
creation-date: 2022-07-12

> last-updated: 2022-07-12

Reviewer 7: This should be updated.

Author: comment incorporated

- Complete end user documentation.
- UTs and e2e tests are present.
- Gather feedback from the users.

#### Tech Preview -> GA
N/A. This feature is for Tech Preview, until decided for GA.

#### Removing a deprecated feature

### Upgrade / Downgrade Strategy

On upgrade:
- Scenario 1: Upgrade to OpenShift version having support for adding tags.
  The new status field won't be populated since it is only populated by the
  installer and that can't have happened if the cluster was installed from a
  prior version. Components that consume the new field should take no action
  since there won't be any tags to apply.
- Scenario 2: Upgrade from OpenShift version having support for adding tags to higher:
  Cluster operators that add tags to Azure resources created for the cluster
  should refer to the tag fields and take action. For any new resource created
  post-upgrade, the operator managing the resource will add the user-defined tags
  to the resource. But the same does not apply to already existing resources,
  components may or may not update the resources with the user defined tags.

On downgrade:
- Scenario 1: Cluster installed with OpenShift version having support for adding tags

>   and later downgrading to a lower version not having support for addding tags.

Reviewer 7: Typo: "addding".

Is it possible to install one version and then downgrade to an earlier version?

Author: comment incorporated.

  The new status field won't be populated since it is only populated by the
  installer and that can't have happened if the cluster was installed from a
  earlier version and upgraded to version having support for tags.
- Scenario 2: Downgrading a cluster to installed OpenShift version not having support 

>   for adding tags from an OpenShift version having support for addding tags.

Reviewer 7: The difference between the two scenarios is unclear.  Both seem to describe downgrading _from_ the same version:

* Scenario 1: "Cluster installed with OpenShift version having support for adding tags".
* Scenario 2: "Downgrading a cluster [...] from an OpenShift version having support for addding tags".

And both scenarios seem to describe downgrading _to_ the same version:

* Scenario 1: "Cluster [...] downgrading to a lower version not having support for addding tags".
* Scenario 2: "Downgrading a cluster to installed OpenShift version not having support for adding tags".

So let's say the feature is introduced in 4.13.  Are both scenarios describing 4.13→4.12?  In both cases, the "installed" version has support for adding tags, so you aren't describing 4.12→4.13→4.12, right?  

It might be possible to simplify this section.  I don't know that you need to describe every possible scenario; usually I just describe what happens if the feature is introduced in 4._n_ and the cluster admin downgrades from 4._n_ to 4.(_n_-1).  Is this description more detailed in order to capture some nuance that is particular to the EP?

Author: Yes, it is possible using allow-explicit-upgrade and force options, but it is not suggested.
Yes, I agree, I have simplified this section with just the scenarios that are important and practical.

  )
}

resource "azurerm_resource_group" "main" {
  tags     = local.tags
}
```

The tag of the form `kubernetes.io_cluster.<cluster_id>:owned` added by openshift, where
cluster_id is a string generated by concatenating user inputted cluster name and a random 
string will be limited to a maximum length of 27 characters by trimming long cluster name 
to 21 characters.

#### Caveats
1. Updating or removing resource tags added by Openshift, using an external interface
   may or may not be reconciled by the operator managing the resource.
2. Updating tags of individual resources is not supported and any tag present in 
   `.status.platformStatus.azure.resourceTags` of `infrastructure.config.openshift.io/v1` resource 
   will result in adding tags to all Openshift managed Azure resources. 

### Risks and Mitigations

### Drawbacks
- User-defined tags cannot be updated on an Azure resource which is not managed by an
  operator. In this proposal, the changes proposed and developed will be part of
  openshift-* namespace. External operators are not in the scope.
  User-defined tags can be updated on the following Azure resources.
  - ResourceGroup
  - Storage Account
  - DNS Zones

>   - DNS Records

Reviewer 8: I went through [Microsoft's docs](https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/tag-support) to double-check, and it wasn't clear to me that "DNS records" had tag support (e.g. it says that dnszones/A records do not, privateDnsZones /A records do not, etc.). Can you confirm that tags are supported?

Author: Yes, these resources do not have explicit tag field, but it's added as metadata for these resources.

  - ApplicationSecurityGroup
  - AvailabilitySet
  - Group
  - LoadBalancer
  - PublicIPAddress
  - RouteTable
  - SecurityGroup
  - VirtualMachineExtension
  - NetworkInterface
  - VirtualMachine
  - VirtualNetwork

- Because tags in `userTags` must be applied to all Azure resources that OpenShift creates,
  tags must fit all the constraints of the various Azure resources as described below:
    1. Tag names cannot have `microsoft`, `azure`, `windows` prefixes which are
       reserved for Azure use.
    2. An Azure storage account has a limit of 128 characters for the tag name.

>     3. An Azure DNS zone or Traffic or Front Door resource tag name cannot have spaces,

Reviewer 8: According to [Microsoft's docs](https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/tag-support), "Front Door doesn't support the use of # or : in the tag name."

Author: Yes, tag names cannot contain those characters and is restricted in OCP as well. It's captured [here](https://github.com/openshift/enhancements/pull/1191/files/47f5b58a740b0f8a16c98baf5af899b7555c59a4#diff-07e0630cc477ba74bc69ea6a15ecefd4f5ce32927210e4e7b0b89d423d2193dbR78).
